Performance tweaks

edding3000 opened the pull request:
<!-- Thank you for contributing to OWASP ModSecurity, your effort is greatly appreciated -->
<!-- Please help us by adding the information below in this PR so it aids reviewers -->

## what

Range based for loops with references are already used in this project, but in a few places not.
I am not sure why. I changed this code locations and executed unit tests.
Also use '*' for pointers when using auto keyword, makes it more readable.

<!--
- Describe high-level what changed as a result of these commits (i.e. in plain-english, what do these changes mean?)
- Use bullet points to be concise and to the point.
-->

## why

Improve performance a little bit.

<!--
- Provide the justifications for the changes (e.g. business case).
- Describe why these changes were made (e.g. why do these commits fix the problem?)
- Use bullet points to be concise and to the point.
-->

## questions

"const auto&" could also be used in many places. Is there a reason why it is not used?

edding3000 committed:
Use reference in range-based for loop

edding3000 committed:
Use explicit '*' with auto for pointers

airween commented on src/engine/lua.cc:
```diff
@@ -297,7 +297,7 @@ int Lua::getvars(lua_State *L) {
 
     lua_newtable(L);
-    for (auto i : l) {
+    for (auto* i : l) {
```
```suggestion
    for (const auto* i : l) {
```
The [original](https://github.com/owasp-modsecurity/ModSecurity/commit/a676f313c3f9277ae007b09d79fa698f61947ce6) code was written about 8 years ago, the author is no longer involved in development.

If I had to guess why they used value copy mechanism, I would say they wanted to avoid any chance to modify the value on Lua stack (except for intentional change with `setvar()`).

If we want to keep this restriction, please consider to use `const` modifier here.

airween commented on src/rules_exceptions.cc:
```diff
@@ -195,7 +195,7 @@ bool RulesExceptions::contains(int a) {
     }
 
-    for (auto z : m_ranges) {
+    for (auto& z : m_ranges) {
```
```suggestion
    for (const auto& z : m_ranges) {
```
Please see `cppcheck`'s [warning](https://github.com/owasp-modsecurity/ModSecurity/actions/runs/19540276744/job/55952868889?pr=3468#step:5:3289):
```
warning: src/rules_exceptions.cc,198,style,constVariableReference,Variable 'z' can be declared as reference to const
```

airween replied:
Sonarcloud also reports [this](https://sonarcloud.io/project/issues?id=owasp-modsecurity_ModSecurity&pullRequest=3468&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true).

edding3000 replied:
done

airween commented on src/rules_exceptions.cc:
```diff
@@ -212,7 +212,7 @@ bool RulesExceptions::merge(RulesExceptions *from) {
         }
     }
-    for (auto b : from->m_ranges) {
+    for (auto& b : from->m_ranges) {
```
```suggestion
    for (const auto& b : from->m_ranges) {
```
Please see cppcheck's [warning](https://github.com/owasp-modsecurity/ModSecurity/actions/runs/19540276744/job/55952868889?pr=3468#step:5:3290):
```
warning: src/rules_exceptions.cc,215,style,constVariableReference,Variable 'b' can be declared as reference to const
```

edding3000 replied:
done

airween commented on src/transaction.cc:
```diff
@@ -1700,7 +1700,7 @@ std::string Transaction::toJSON(int parts) {
             strlen("messages"));
         yajl_gen_array_open(g);
-        for (auto a : m_rulesMessages) {
+        for (auto& a : m_rulesMessages) {
```
```suggestion
        for (const auto& a : m_rulesMessages) {
```
Probably here you could use `const` too.

edding3000 replied:
done

airween commented on src/transaction.cc:
```diff
@@ -1721,7 +1721,7 @@ std::string Transaction::toJSON(int parts) {
                 strlen("tags"));
             yajl_gen_array_open(g);
-            for (auto b : a.m_tags) {
+            for (auto& b : a.m_tags) {
```
```suggestion
            for (const auto& b : a.m_tags) {
```
Probably here you could use `const` too.

edding3000 replied:
done

edding3000 commented:
I was not sure if there was any restriction not to use const, thats why i asked. I could extend the PR with more positions for "const auto&" if you like.

airween commented:
> I was not sure if there was any restriction not to use const, thats why i asked. I could extend the PR with more positions for "const auto&" if you like.

Sure, go! If the tests will be passed, I'll be happy :).

Btw you could review the tests too, may be you will find some lack there too (for eg. you're checking the code...)

edding3000 committed:
Add const modifier to range-based for loops

edding3000 committed:
Use const pointers

edding3000 committed:
Use explicit '*' with auto for pointers in tests

edding3000 committed:
Remove c-cast from test

edding3000 committed:
Switch to const string reference as nothing is modified

edding3000 committed:
Add const to parameter

with changes from e9bf9ad0e2c4043dc68da276dcc0109e47b066ab nothing is modified anymore

airween commented:
Hi @edding3000,

sorry for the long delay, it seems like this PR is ready to merge. Just one question: could you merge the current v3/master into this one?

airween commented:
@edding3000,

also please take a look at the few request changes.

edding3000 committed:
Merge branch 'v3/master' into performance_tweaks

edding3000 committed:
Fix const correctness warning

edding3000 commented:
@airween I merged v3/master. cppcheck fails now. Can you see the reason?

airween commented:
> @airween I merged v3/master. cppcheck fails now. Can you see the reason?

Yes:
```
2026-02-17T18:48:51.1404560Z warning: src/run_time_string.cc,54,style,functionConst,The member function 'modsecurity::RunTimeString::evaluate' can be made a const function. Making this function 'const' should not cause compiler errors. Even though the function can be made const function technically it may not make sense conceptually. Think about your design and the task of the function first - is it a function that must not change object internal state?
```

I think `cppcheck` realized that you [changed](https://github.com/owasp-modsecurity/ModSecurity/pull/3468/changes#diff-2bd76cacd9e1f795d3b62a65bd3d129d6c3672c38ec902b732fea39f8ad4b551R56) the loop's variable to `const`, and figured out the function can't change the object (I don't see the relationship what this `const` triggered this error), so it expects to make the fn to `const` - but warns us that it might be wrong.

So what you can try now:
* change the function signature to `const`: `std::string RunTimeString::evaluate(Transaction *t, Rule *r) const {`
* if the build will be failed, rollback that change and add an inline suppression to that line: `std::string RunTimeString::evaluate(Transaction *t, Rule *r) { // cppcheck-suppress functionConst`
* in this case, please add this note to the PR's description

Do you have `cppcheck` on your local dev environment? It's important to use the latest release, and on Linux only Debian SID contains that version. [Here](https://github.com/owasp-modsecurity/ModSecurity/blob/v2/master/.github/workflows/ci.yml#L141-L176) you can see how do we use it in Docker.

edding3000 commented:
Thx for your hint. I build cppcheck from source, but it did not find this - perhaps I did not use correct settings. Anyway, warning is fixed now.

airween commented on headers/modsecurity/rules_exceptions.h:
```diff
@@ -51,7 +51,7 @@ class RulesExceptions {
     bool addRange(int a, int b);
     bool addNumber(int a);
     bool contains(int a);
```
@edding3000 I think this suggestion makes sense.

airween replied:
Same as above, this is public also.

airween commented on src/run_time_string.h:
```diff
@@ -48,9 +48,9 @@ class RunTimeString {
-    std::string evaluate(Transaction *t);
-    std::string evaluate(Transaction *t, Rule *r);
-    std::string evaluate() {
+    std::string evaluate(Transaction *t) const;
+    std::string evaluate(Transaction *t, Rule *r) const;
+    std::string evaluate() const{
         return evaluate(NULL);
     }
```
This one also makes sense.

airween commented on headers/modsecurity/rule_with_actions.h:
```diff
@@ -80,7 +80,7 @@ class RuleWithActions : public Rule {
 
     std::vector<actions::Action *> getActionsByName(const std::string& name,
         const Transaction *t);
```
This suggestion makes sense.

airween replied:
Just for sure I took a second review.

Because of this header file is part of the public API, and `containsTag` appears between the symbols, this can cause to an unresolved symbol error.
```bash
$ nm -D /path/to/libmodsecurity.so.3 | grep containsTag
0000000000276cc6 T _ZN11modsecurity15RuleWithActions11containsTagERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPNS_11TransactionE
```

airween reviewed (changes requested):
@edding3000 thanks, please take a look at Copilot reviews and my suggestions below.

airween commented:
@edding3000,

one more thing: if you will align those suggested modifications (make wrappers to functions which changed their signatures), could you take a look at [this](https://github.com/owasp-modsecurity/ModSecurity/blob/v3/master/src/variables/variable.h#L157) too?
```C++

    bool toOmit(std::string a) {
        for (auto &z : *this) {
            if (z->match(a)) {
                return true;
            }
        }
        return false;
    }
```
This can be a `const` argument too, it seems like the newest `cppcheck` [warns](https://github.com/owasp-modsecurity/ModSecurity/pull/3504#issuecomment-4019424800) about that.
